slurm_openstack_tools/reboot.py

```diff
@@ -1,160 +1,122 @@
+    return True
+
+
+def reboot_node(conn, server_id, reboot_type="HARD"):
+    """
+    Reboot the node (default: hard reboot).
+    """
+    logger.info(f"Rebooting server {server_id} with {reboot_type.lower()} reboot")
+    conn.compute.reboot_server(server_id, reboot_type=reboot_type)
+    return True
+
+
+def process_node(conn, node):
+    """
+    Process a single node by comparing its target and current images.
+    """
+    hostvars = read_hostvars(node)
+    if not hostvars:
+        logger.info(f"No hostvars defined for node {node}, skipping...")
+        return
+
+    server_id = hostvars.get("instance_id")
+    target_image_id = hostvars.get("image_id")
+
+    if not server_id:
+        logger.info(f"Node {node} is not an OpenStack node, ignoring...")
+        return
+
+    if not target_image_id:
+        logger.info(f"Node {node} does not have a target image defined, performing reboot...")
+        reboot_node(conn, server_id)
+        return
+
+    current_image_id = get_current_image(conn, server_id)
+    if current_image_id != target_image_id:
+        logger.info(f"Node {node} requires rebuild: current image {current_image_id}, target image {target_image_id}")
+        rebuild_node(conn, server_id, target_image_id)
+    else:
+        logger.info(f"Node {node} is already using the target image, performing reboot...")
+        reboot_node(conn, server_id)
 
-    If the reason starts with "rebuild" then the node is rebuilt; arguments to
-    `openstack.compute.rebuild_server()` may optionally be passed by including
-    space-separated `name:value` pairs in the reason.
 
-    If the reason does not start with "rebuild" then the node is rebooted.
-    Note the "reason" message must be MAX_REASON_LENGTH or less.
+def main():
+    """
+    Main function to process nodes from the Slurm-provided hostlist.
+    """
+    if len(sys.argv) < 2:
+        logger.error("Usage: <script> <hostlist>")
+        sys.exit(1)
 
-    Messages and errors are logged to syslog.
+    hostlist = sys.argv[1].split(",")
+    conn = openstack.connection.from_config()
 
-    Requires:
-    - Python 3 with openstacksdk module
-    - The node's Openstack ID to have been set by cloud init in
-      `/var/lib/cloud/data/instance-id`
-    - An application credential:
-        - with at least POST rights to /v3/servers/{server_id}/action
-        - available via a clouds.yaml file containing only one cloud
-    """ # noqa E501
-    server_uuid = get_openstack_server_id()
-    if not server_uuid:
-        logger.info("rebooting non openstack server")
-        do_reboot()
-    else:
-        reason = get_reboot_reason()
-        if not reason.startswith("rebuild"):
-            logger.info("rebooting openstack server, locally")
-            do_reboot()
-        else:
-            logger.info("rebuilding openstack server")
-            rebuild_openstack_server(server_uuid, reason)
+    for node in hostlist:
+        process_node(conn, node)
```

Comment:
How should we deal with errors here, I think we probably need to Log here if there were any errors, with a try / catch, and set a raise_error = true flag if we hit any errors in the loop, then do exit with -1 if we hit some error?